pyro/distributions/ordered_logistic.py

```diff
@@ -12,38 +12,45 @@ class OrderedLogistic(Categorical):
     accepts a vector of ``cutpoints`` which are an ordered vector of real numbers
     denoting baseline cumulative log-odds of the individual categories, and a
-    model vector ``predictors`` which modifies the baselines for each sample
+    model vector ``predictor`` which modifies the baselines for each sample
     individually.
 
     These cumulative log-odds are then transformed into a discrete cumulative
     probability distribution, that is finally differenced to return the probability
     mass matrix ``p`` that specifies the categorical distribution.
 
-    :param Tensor predictors: A tensor of predictor variables of arbitrary
+    :param Tensor predictor: A tensor of predictor variables of arbitrary
         shape. The output shape of non-batched samples from this distribution will
-        be the same shape as ``predictors``.
+        be the same shape as ``predictor``.
     :param Tensor cutpoints: A tensor of cutpoints that are used to determine the
-        cumulative probability of each entry in ``predictors`` belonging to a
+        cumulative probability of each entry in ``predictor`` belonging to a
         given category. The first `cutpoints.ndim-1` dimensions must be
-        broadcastable to ``predictors``, and the -1 dimension is monotonically
+        broadcastable to ``predictor``, and the -1 dimension is monotonically
         increasing.
     """
 
     arg_constraints = {
-        "predictors": constraints.real,
+        "predictor": constraints.real,
         "cutpoints": constraints.ordered_vector,
     }
 
-    def __init__(self, predictors, cutpoints, validate_args=None):
-        self.predictors = predictors
+    def __init__(self, predictor, cutpoints, validate_args=None):
+        self.predictor = predictor
         self.cutpoints = cutpoints
-        # get shape for input to Categorical dist
-        p_shape = predictors.shape + (cutpoints.shape[-1] + 1,)
         # calculate cumulative probability for each predictor
-        q = torch.sigmoid(cutpoints - predictors.view(predictors.shape + (1,)))
+        q = torch.sigmoid(cutpoints - predictor.view(predictor.shape + (1,)))
         # turn cumulative probabilities into probability mass of categories
-        p = torch.zeros(p_shape)
+        p_shape = q.shape[:-1] + (q.shape[-1] + 1,)
+        p = torch.zeros(p_shape, dtype=q.dtype, device=q.device)
         p[..., 0] = q[..., 0]
         p[..., 1:-1] = (q - torch.roll(q, 1, dims=-1))[..., 1:]
         p[..., -1] = 1 - q[..., -1]
         super(OrderedLogistic, self).__init__(p, validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        batch_shape = torch.Size(batch_shape)
+        new = self._get_checked_instance(OrderedLogistic, _instance)
+        predictor = self.predictor.expand(batch_shape)
+        cutpoints = self.cutpoints.expand(batch_shape + (self.cutpoints.shape[-1],))
+        new.__init__(predictor, cutpoints)
+        return new
```

Comment:
Hmm, the purpose of `expand` is to avoid recomputing the logics in the constructor again. I think it is better to do
```python
new.predictor = ...
new.cutpoints = ...
return super(OrderedLogistic, self).expand(batch_shape, new)
# or return Categorical.expand(self, batch_shape, new)
```
If you broadcast `predictor` and `cutpoints` here, then could you also do so in `__init__`?, kind of
```python
super(OrderedLogistic, self).__init__(p, validate_args=validate_args)
self.predictor = predictor.expand(self.batch_shape)
self.cutpoints = cutpoints.expand(self.batch_shape + cutpoints.shape[-1:])
```

Reply:
nit: you missed a new line at the end of this file

Reply:
ahh yes, that makes sense